### PR TITLE
dependencies: Hide internals

### PR DESCRIPTION
**Pull request by efritz**
Hide lockfiles and store implementation from outside consumers. The `Service` struct itself should be what all external consumers interact with directly. Non-internal packages (http/graphql transports and background jobs) will be located outside of the service's internal package.

This is a pre-step to extending the dependencies service with background jobs and its own transport layers.

**Please review by commit**:

Hopefully review-by-commit will make this short-work for readers. Please pay special attention to the two marked commits as I believe the implementation is in alignment with what was originally _wanted_, but would like to double-check. I'm happy to respond to readability concerns in that code or the related tests as well.

- 30807f13 Moves the `store` and `lockfiles` behind an `internal` package so that they cannot be imported outside of the dependencies service root.
- 92fb4bd0 Add new methods to the service that delegates to the store. We need to do type wrapping conversions here (unfortunately; for now) because go-mockgen is not playing nicely with the private package + type aliases. I'll file an issue and make sure we can shed some layers here later if we're not doing anything useful with them by that time.
- 0027042e Updates users in a way that is obviously behavior-preserving.
- c58bf5f3 Removes an alias package and its single user.
- 5868a5f5 **IMPORTANT** Updates go, jvm, and and npm extsvcs to use a dependencies service rather than its store
- 3a559f93 **IMPORTANT** Update tests in `internal/repos` to not depend on an actual database
- ... mockgen, chores, responses to feedback, etc

## Test plan

Existing unit / integration tests. Tested dependency searches locally and [main-dry-run](https://buildkite.com/sourcegraph/sourcegraph/builds/143375) passed.

### CODE REVIEW

File: internal/codeintel/dependencies/service.go [hunk 1]
```diff
@@ -218,3 +218,50 @@ func constructLogFields(repoRevs map[api.RepoName]types.RevSpecSet) []log.Field
...
+func (s *Service) UpsertDependencyRepos(ctx context.Context, deps []DependencyRepo) ([]DependencyRepo, error) {
+	drs, err := s.dependenciesStore.UpsertDependencyRepos(ctx, to(deps))
+	if err != nil {
+		return nil, err
+	}
+
+	return from(drs), nil
+}
+
+func to(local []DependencyRepo) []store.DependencyRepo {
```
Comment (tsenart) [inline internal/codeintel/dependencies/service.go:251]: Would it be useful to re-use a `shared.DependencyRepo` type to avoid these conversions?

File: internal/repos/iface.go [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package repos
+
+import (
+	"context"
+
+	"github.com/sourcegraph/sourcegraph/internal/codeintel/dependencies"
+)
+
+type DependenciesService interface {
+	ListDependencyRepos(ctx context.Context, opts dependencies.ListDependencyReposOpts) ([]dependencies.DependencyRepo, error)
```
Comment (tsenart) [inline internal/repos/iface.go:10]: Is this needed to avoid a circular import? If so, can we add a comment saying so?

File: internal/repos/sources.go [hunk 3]
```diff
@@ -92,15 +93,17 @@ type RepoGetter interface {
 
 type DBSource interface {
```
Comment (tsenart) [inline internal/repos/sources.go:93]: This should no longer be called DBSource since it has nothing to do with DBs. We should probably instantiate the service outside and then have a `WithDependenciesService()` option passed.

We can avoid exporting this type by using an anonymous interface inside the `WithDependenciesService` option.

File: internal/repos/npm_packages_test.go [hunk 4]
```diff
@@ -79,12 +81,64 @@ func TestGetNpmDependencyRepos(t *testing.T) {
 	}
 }
 
-func setupDependenciesInDB(t *testing.T) (*sql.DB, *dependenciesStore.Store, context.Context, []string) {
+func setupDependenciesService(t *testing.T, drs []dependencies.DependencyRepo) (context.Context, DependenciesService) {
 	t.Helper()
-	db := dbtest.NewDB(t)
-	store := dependenciesStore.TestStore(database.NewDB(db))
 	ctx := context.Background()
```
Comment (tsenart) [inline internal/repos/npm_packages_test.go:79]: nit: Returning this context from this func seems orthogonal. I'd return only the DependenciesService.

File: internal/repos/npm_packages_test.go [hunk 5]
```diff
@@ -79,12 +81,64 @@ func TestGetNpmDependencyRepos(t *testing.T) {
...
+	depsSvc := NewMockDependenciesService()
+
+	depsSvc.ListDependencyReposFunc.SetDefaultHook(func(ctx context.Context, opts dependencies.ListDependencyReposOpts) (matching []dependencies.DependencyRepo, _ error) {
```
Comment (tsenart) [inline internal/repos/npm_packages_test.go:82]: We effectively converted this from an integration test to a unit test. Is this change necessary?

File: internal/repos/npm_packages_test.go [hunk 6]
```diff
@@ -79,12 +81,64 @@ func TestGetNpmDependencyRepos(t *testing.T) {
...
+			matching = matching[:opts.Limit]
+		}
+
+		return matching, nil
+	})
+
+	return ctx, depsSvc
+}
 
+func parseDepStrs(t *testing.T, depStrs []string) []dependencies.DependencyRepo {
```
Comment (tsenart) [inline internal/repos/npm_packages_test.go]: nit: More ergonomic to read and write with `depStrs ...string)`

File: internal/repos/npm_packages_test.go [hunk 7]
```diff
@@ -79,12 +81,64 @@ func TestGetNpmDependencyRepos(t *testing.T) {
...
+	parsedDependencyRepos := []dependencies.DependencyRepo{}
```
Comment (tsenart) [inline internal/repos/npm_packages_test.go]: nit: My Goland complains about this and wants us to use the zero value since in some cases avoids allocations.

```
var parseDependencyRepos []dependencies.DependencyRepo
```


Another thing: `dependencies.Repo` would read better than `dependencies.DependencyRepo`

File: internal/codeintel/dependencies/shim/git.go [hunk 8]
```diff
@@ -0,0 +1,34 @@
+package shim
```
Comment (tsenart) [inline internal/codeintel/dependencies/shim/git.go]: Can you add a package comment explaining what this does and why it exists? What's a shim in this context, why we need it, etc.

File: internal/codeintel/dependencies/live/init.go [hunk 9]
```diff
@@ -0,0 +1,28 @@
+package live
```
Comment (tsenart) [inline internal/codeintel/dependencies/live/init.go:1]: So, shim became live, but still not clear why this package exists and the reasoning for its name.

File: internal/codeintel/dependencies/live/init.go [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+
+func GetService(db database.DB, syncer dependencies.Syncer) *dependencies.Service {
+	return dependencies.GetService(db, NewGitService(db), syncer)
+}
+
+// GetServiceWithoutSyncer should be used when constructing a dependencies service that only uses the
+// system-level behaviors (e.g., listing repositories to sync since the last request). This syncer value
+// will issue errors on invocation indicating that gitserver/repoupdater services are not expected to
+// invoke such methods.
+func GetServiceWithoutSyncer(db database.DB) *dependencies.Service {
```
Comment (tsenart) [inline internal/codeintel/dependencies/live/init.go:20]: Seems a bit overkill to have two methods. How about just passing `nil` to the syncer argument? Then where it's used we check if it's nil and only call methods on it if it's not.

**Review by tsenart — COMMENTED** [review]
Reviewing commit by commit means there are some redundant comments :)

File: internal/repos/npm_packages_test.go (hunk 4 above)
Comment (efritz) [reply internal/repos/npm_packages_test.go:79]: Updated (but was an unrelated change to this PR).

Comment (efritz) [reply internal/repos/npm_packages_test.go]: 1a56dffdce034bd9792de0082a952c4c1cedf166

File: internal/repos/npm_packages_test.go (hunk 6 above)
Comment (efritz) [reply internal/repos/npm_packages_test.go]: No longer relevant after later commits.

File: internal/codeintel/dependencies/live/init.go (hunk 10 above)
Comment (efritz) [reply internal/codeintel/dependencies/live/init.go]: c267f1b9b3daf6f394f2165b6a838da4efb8f099

File: internal/codeintel/dependencies/live/init.go (hunk 9 above)
Comment (efritz) [reply internal/codeintel/dependencies/live/init.go:5]: f519d901f0895edb8bca53be097e233b1cddb15e

File: internal/codeintel/dependencies/service.go (hunk 1 above)
Comment (efritz) [reply internal/codeintel/dependencies/service.go]: 4f04b73f66adabe1e8b3b4774f1d537a002c6197

File: internal/repos/npm_packages_test.go (hunk 7 above)
Comment (efritz) [reply internal/repos/npm_packages_test.go]: 4f04b73f66adabe1e8b3b4774f1d537a002c6197

File: internal/repos/sources.go (hunk 3 above)
Comment (efritz) [reply internal/repos/sources.go]: I swore I did actually update this at some point in this journey...

3635ca13d81dadc43bbb1f2af1bb74abd2888dcb

File: internal/repos/npm_packages_test.go (hunk 5 above)
Comment (efritz) [reply internal/repos/npm_packages_test.go]: e0aa3be36d02155a39f1a36ecea278e097c8374f and e0aa3be36d02155a39f1a36ecea278e097c8374f

File: internal/repos/iface.go (hunk 2 above)
Comment (efritz) [reply internal/repos/iface.go]: 3c67cc5bb72a8a5baf56d4d17321d689890e1c96 and e0aa3be36d02155a39f1a36ecea278e097c8374f

File: cmd/gitserver/server/vcs_syncer_go_modules.go [hunk 11]
```diff
@@ -37,16 +37,16 @@ var placeholderGoDependency = func() *reposource.GoDependency {
 // from Go module proxies and converting them to Git repositories.
 type GoModulesSyncer struct {
 	connection *schema.GoModulesConnection
-	depsStore  repos.DependenciesStore
+	depsSvc    repos.DependenciesService
```
[the remark this answers is not in the document]
Comment (efritz) [reply cmd/gitserver/server/vcs_syncer_go_modules.go]: 3c67cc5bb72a8a5baf56d4d17321d689890e1c96 and e0aa3be36d02155a39f1a36ecea278e097c8374f

**Comment by efritz** [comment]
@tsenart Responded to all your feedback.

image

**Review by tsenart — APPROVED** [review]
:shipit:
